### ISSUE TITLE
Drawing issue on Mac OS

### ISSUE THREAD

**Issue by Jakob-SD** [kevinconway/remouseable#8]
Hi there

I have a funky issue on Mac OS. When I start drawing, the drawing application only intercepts the point where I start the drawing and when I lift the pen again. Everything in between is not drawn. The result is that circles shows as small straight lines.

I have attached a screen recording here
https://youtu.be/J0pd1F-aPnQ

Thanks in advance for any help

[EDIT]
I have tried several drawing applications with same result

**Comment by kevinconway** [comment]
I fully updated my reMarkable 1 tablet and was not able to recreate the issue on any OS but did have another person report this same issue for OSX over email. Do you have a reMarkable 2?

If you are comfortable building the project then check out check and run the `debug-v2-events` branch. There is an extra flag there called `--debug-events` that will switch the application from driving the mouse to emitting JSON events. If you run `go run main.go --debug-events > debug.json` and upload the JSON file then I can help see if here is a change in the way the device events are captured that is causing the issue. The output is generated quickly so it's best to start the app in debug mode, draw a shape, and then ctrl+c.

**Comment by Jakob-SD** [comment]
Thank you very much for your reply. Yes I have the ReMarkable 2 and spent quite a few hours yesterday on testing. It seem to depend on the drawing app. 
Onenote and Notability has this issue while Sketchpad works as expected. 

I have tested with several other drawing apps with mixed results. 

I will try to follow your guide, and return back with the result 👍

**Comment by kevinconway** [comment]
If it varies by drawing app then it may mean that certain apps are interpreting the mouse events differently. Currently, when you touch the pen to the tablet remousable issues a single `MOUSE_DOWN` event until the pen is lifted when it then issues a `MOUSE_UP`. The result in most applications is that the system interprets that as holding down the mouse button. For example, you can draw a diagonal line on your desktop to select files because it's identical to click and drag. It sounds like the reMarkable tablet events are unchanged so the debug events may not be as helpful as I hoped.

What would be surprising to me is if you drew with the mouse in Onenote and got a different result than the than the tablet since click and drag _should_ be exactly the same given that remousable is acting as your mouse.

**Comment by Jakob-SD** [comment]
Actually drawing with the mouse works. These are the apps I have tested so far

- OneNote, works with mouse but not with remouseable
- Notability, works with mouse but not with remouseable
- InkScape, works with mouse but not with remouseable
- Affinity Designer, works with mouse and remouseable
- Sketchpad, works with mouse and remouseable
- draw.io, works with mouse and remouseable

The funny thing is that I tested with Onenote and Notability first, as they were my preferred tools.

**Comment by kevinconway** [comment]
tl;dr If you have xcode installed on your OSX machine, clone <https://github.com/kevinconway/remouseable/tree/osx-drag> and `go run -a main.go --osx-drag-enabled`. This fixed the issue for me but I'd like to know if it fixes the issue for you as well. If it works then I'll likely make it the default option since it appears to be backwards compatible for all use cases.


details:

I had some time to test on OSX today and think I have a lead. It turns out that OSX has a special mouse event for drag that is different than a regular move whereas other systems only have move and imply drag when the mouse is held down. If this turns out to be the issue then I imagine it went undetected for a while because most web apps only look at the same CLICK -> MOVE -> UNCLICK events that remousable is sending and only applications optimized to run on OSX work with the custom DRAG event.

I tested with Inkscape because it was the most easily available to me. Drawing on windows was issue free. I have a linux machine I can use to test later but I suspect that linux is also issue free. OSX would let me draw a line or two but would then crash. Swapping in the OSX drag event helped but it made moving the mouse lag behind my pen by a very large amount.

To compare, I enabled the drag event on windows where drag is an alias for move and move is what it was using before. Surprisingly, the same lag was introduced. This made me think there's something wrong in my code as it relates to converting move to drag that is introducing the latency. It was almost like the move is being registered as a "smooth move" that has an animation and cooldown that causes it to queue up events and lag. It turns out that's exactly what was happening. I found a 10ms sleep in the drag function that I copied from <https://github.com/go-vgo/robotgo>. I removed the sleep and things worked as expected on all platforms.

Inkscape still seems a bit buggy and crashes occasionally when drawing. It seems to be more of an Inkscape issue than a mouse one, though. All of the drawing canvas apps I have access to seem to work with <https://github.com/kevinconway/remouseable/tree/osx-drag>. I'm curious if you see any bugs using Onenote with the work-in-progress branch.

**Comment by Jakob-SD** [comment]
I got the branch but you lost me with "go run -a main.go --osx-drag-enabled". :)
Can you try to explain again?

I have xcode installed as a part of other projects

**Comment by Jakob-SD** [comment]
Ok, I think I managed to following your guide. I installed go with Brew and ran 

    go run -a main.go --osx-drag-enabled --event-file /dev/input/event1 --ssh-password *******

It works as expected in OneNote now, so perhaps you fixed the issue. Thank you very much for your effort on this matter. Will you release a new binary or should I use the previous command?

**Comment by kevinconway** [comment]
Ah, yeah. Building the project also requires that you have [Go](https://golang.org/) installed. Go installs a command called `go` that includes the compiler. From inside the repository you use `go run <filename>` to build and execute something. The typical executable file name in go `main.go` and there's one of those in the root of the repository. The `-a` flag forces the compiler to ignore cache and rebuild which I found necessary sometimes because Go caches the compilation of the C code that is embedded in the project. `go run -a main.go` from the repository root will rebuild and execute remousable. You can add flags to the execution just like the pre-compiled executables in the Github release. `--osx-drag-enabled` is a flag I added to that new branch that switches the remousable behavior from issuing MOVE events on OSX to DRAG events which should work better. `go run -a main.go --ssh-password="YOUR PASSWORD" --osx-drag-enabled` will run everything with the new behavior for testing.

I typed ^^ before seeing that you got it running and figured I'd leave it there in case it still helped explain anything.

> It works as expected in OneNote now, so perhaps you fixed the issue. Thank you very much for your effort on this matter. Will you release a new binary or should I use the previous command?

Good news! I'll publish a new version when I get a chance. I need to tweak the code to make DRAG the default and then rebuild it on the different operating systems. It might take a few days for me to sort that out so I'd suggest you continue to run it from source in the meantime. Thanks for trying going through all the build steps and trying out the branch with me. I'll leave this issue open until the new release is out. You should get an email when I close it to let you know there's a new version available.

**Comment by Jakob-SD** [comment]
Sound awesome. I have been testing some more, and only found one thing different from using a mouse. In Notability I can draw a line and keep the mouse pressed for a few ms at the end of the line. That makes the line straight. It is not an issue at all, so just to let you know.

I wonder if I can build the binary myself 🤔

**Comment by kevinconway** [comment]
You can build the binary with `go build -o remousable main.go` where `-o` is the name of the file it will create.

**Comment by kevinconway** [comment]
New release is available at https://github.com/kevinconway/remouseable/releases/tag/v0.4.0 . Thanks for opening the issue and testing things out with me. Open another issue if the release binaries don't work as expected.
